File: 算法/C/第3章 排序算法/3.9 基数排序/radix_sort.c
#include <stdio.h>
#include <stdlib.h>
#include "array.h"

void showArr(int *arr, int n) {
    for(int i = 0; i < n; i++) {
        printf("%d ", arr[i]);
    }
    printf("\n");
}

void radixSort(int *arr, int n) {
    int max = arr[0];
    for(int i = 1; i < n; i++) {
        if(arr[i] > max) {
            max = arr[i];
        }
    }

    // 创建桶
    int bucketNum = 10;
    Array *buckets[bucketNum];
    for(int i = 0; i < bucketNum; i++) {
        buckets[i] = initArray();
    }

    // 从个位开始
    for(int exp = 1; max / exp > 0; exp *= 10) {
        // 把元素放到对应桶中
        for(int i = 0; i < n; i++) {
            int num = (arr[i] / exp) % 10;
            append(buckets[num], arr[i]);
        }

        // 按顺序取出元素
        int cnt = 0;
        for(int i = 0; i < bucketNum; i++) {
            for(int j = 0; j < buckets[i]->size; j++) {
                arr[cnt++] = buckets[i]->data[j];
            }
            clearArray(buckets[i]);
        }
    }

    for(int i = 0; i < bucketNum; i++) {
        deleteArray(buckets[i]);
    }
}

int main() {
    int arr[] = {63, 157, 189, 51, 101, 47, 141, 121, 157, 156,
                 194, 117, 98, 139, 67, 133, 181, 12, 28, 0, 109};
    int n = sizeof(arr) / sizeof(arr[0]);
    radixSort(arr, n);
    showArr(arr, n);
    return 0;
}

File: 算法/Python/第3章 排序算法/3.9 基数排序/radix_sort.py
def radix_sort(lst):
    n = len(lst)
    max_elem = max(lst)

    # 创建桶
    bucket_num = 10

    # 从个位开始
    exp = 1
    while max_elem // exp > 0:
        # 创建桶
        buckets = [[] for _ in range(bucket_num)]

        # 把元素放到对应桶中
        for i in range(n):
            num = (lst[i] // exp) % 10
            buckets[num].append(lst[i])
        
        # 按顺序取出元素
        cnt = 0
        for bucket in buckets:
            for data in bucket:
                lst[cnt] = data
                cnt += 1

        exp *= 10

def main():
    lst = [63, 157, 189, 51, 101, 47, 141, 121, 157, 156,
           194, 117, 98, 139, 67, 133, 181, 12, 28, 0, 109]
    radix_sort(lst)
    print(lst)

if __name__ == "__main__":
    main()